### PR TITLE
ADEN-4227 | add renderEnded event listener  to enable BTF queue

### PR DESCRIPTION
**Pull request by bognix**

### CODE REVIEW

**Comment by WikiaTech** [comment]
SonarQube analysis reported 6 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [AdEngine2.js#L187](https://github.com/Wikia/app/blob/7888d238e2d5c5cc0fcaf6c4bce8750f11014e7c/extensions/wikia/AdEngine/js/AdEngine2.js#L187): Change this condition so that it does not always evaluate to "true". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS2583)
1. MAJOR [AdEngine2.js#L12](https://github.com/Wikia/app/blob/7888d238e2d5c5cc0fcaf6c4bce8750f11014e7c/extensions/wikia/AdEngine/js/AdEngine2.js#L12): Function has 9 parameters which is greater than 7 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AExcessiveParameterList)
1. MAJOR [helper.js#L16](https://github.com/Wikia/app/blob/7888d238e2d5c5cc0fcaf6c4bce8750f11014e7c/extensions/wikia/AdEngine/js/provider/gpt/helper.js#L16): Function has 12 parameters which is greater than 7 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AExcessiveParameterList)
1. MAJOR [helper.js#L51](https://github.com/Wikia/app/blob/7888d238e2d5c5cc0fcaf6c4bce8750f11014e7c/extensions/wikia/AdEngine/js/provider/gpt/helper.js#L51): Function has a complexity of 14 which is greater than 10 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AFunctionComplexity)
1. INFO [AdEngine2.js#L58](https://github.com/Wikia/app/blob/7888d238e2d5c5cc0fcaf6c4bce8750f11014e7c/extensions/wikia/AdEngine/js/AdEngine2.js#L58): Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)
1. INFO [AdEngine2.js#L81](https://github.com/Wikia/app/blob/7888d238e2d5c5cc0fcaf6c4bce8750f11014e7c/extensions/wikia/AdEngine/js/AdEngine2.js#L81): Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)

File: extensions/wikia/AdEngine/js/AdEngine2.js [hunk 1]
```diff
@@ -133,6 +133,10 @@ define('ext.wikia.adEngine.adEngine', [
 						eventDispatcher.dispatch('adengine.slot.status', {slotName: slotName, status: 'hop'});
 						tracker.track('hop', adInfo);
 						nextProvider();
+					},
+					renderEnded: function(googleSlot) {
+						log(['renderEnded', provider.name, slotName, googleSlot], 'debug', logGroup);
+						tracker.track('renderEnded', googleSlot);
```
Comment (drets) [inline extensions/wikia/AdEngine/js/AdEngine2.js]: In the different hooks we send `adInfo` (something meaningful) to GA, what do we send here (do we want do send it)?

File: extensions/wikia/AdEngine/js/provider/factoryWikiaGpt.js [hunk 2]
```diff
@@ -33,6 +33,7 @@ define('ext.wikia.adEngine.provider.factory.wikiaGpt', [
 	 * @param {function} [extra.beforeSuccess]  - function to call before calling success
 	 * @param {function} [extra.beforeCollapse] - function to call before calling collapse
 	 * @param {function} [extra.beforeHop]      - function to call before calling hop
+	 * @param {boolean}  [extra.onSlotRendered] - function to call before calling renderEnded
```
Comment (drets) [inline extensions/wikia/AdEngine/js/provider/factoryWikiaGpt.js]: `s/boolean/function/`

File: extensions/wikia/AdEngine/js/slot/adSlot.js [hunk 3]
```diff
@@ -27,6 +27,11 @@ define('ext.wikia.adEngine.slot.adSlot', [
 				callbacks.hop(adInfo);
 			}
 		};
+		slot.renderEnded = function(slot) {
+			if (typeof callbacks.renderEnded === 'function') {
+				callbacks.renderEnded(slot);
+			}
+		};
```
Comment (drets) [inline extensions/wikia/AdEngine/js/slot/adSlot.js]: What do you think about small refactoring?
```js
	function create(name, container, callbacks) {
		function makeHook(fn) {
			return function(data) {
				if (typeof fn === 'function') {
					fn(data);
				}
			};
		}

		return {
			name: name,
			container: container,
			collapse: makeHook(callbacks.collapse),
			hop: makeHook(callbacks.hop),
			renderEnded: makeHook(callbacks.renderEnded),
			success: makeHook(callbacks.success)
		};
	}
```

File: extensions/wikia/AdEngine/js/AdEngine2.js (hunk 1 above)
Comment (bognix) [reply extensions/wikia/AdEngine/js/AdEngine2.js]: you are right. I don't think it's worth sending this info to GA

**Comment by WikiaTech** [comment]
SonarQube analysis reported 6 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [AdEngine2.js#L187](https://github.com/Wikia/app/blob/cb2264d34050aabbfa8ac382d00389c63c50d011/extensions/wikia/AdEngine/js/AdEngine2.js#L187): Change this condition so that it does not always evaluate to "true". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS2583)
1. MAJOR [AdEngine2.js#L12](https://github.com/Wikia/app/blob/cb2264d34050aabbfa8ac382d00389c63c50d011/extensions/wikia/AdEngine/js/AdEngine2.js#L12): Function has 9 parameters which is greater than 7 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AExcessiveParameterList)
1. MAJOR [helper.js#L16](https://github.com/Wikia/app/blob/cb2264d34050aabbfa8ac382d00389c63c50d011/extensions/wikia/AdEngine/js/provider/gpt/helper.js#L16): Function has 12 parameters which is greater than 7 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AExcessiveParameterList)
1. MAJOR [helper.js#L51](https://github.com/Wikia/app/blob/cb2264d34050aabbfa8ac382d00389c63c50d011/extensions/wikia/AdEngine/js/provider/gpt/helper.js#L51): Function has a complexity of 14 which is greater than 10 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AFunctionComplexity)
1. INFO [AdEngine2.js#L58](https://github.com/Wikia/app/blob/cb2264d34050aabbfa8ac382d00389c63c50d011/extensions/wikia/AdEngine/js/AdEngine2.js#L58): Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)
1. INFO [AdEngine2.js#L81](https://github.com/Wikia/app/blob/cb2264d34050aabbfa8ac382d00389c63c50d011/extensions/wikia/AdEngine/js/AdEngine2.js#L81): Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)

**Review by drets — APPROVED** [review]

**Comment by WikiaTech** [comment]
SonarQube analysis reported 6 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [AdEngine2.js#L186](https://github.com/Wikia/app/blob/de9bcd7e5b82ac2b4fc80ec39d13d0c1bb6f6b9c/extensions/wikia/AdEngine/js/AdEngine2.js#L186): Change this condition so that it does not always evaluate to "true". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS2583)
1. MAJOR [AdEngine2.js#L12](https://github.com/Wikia/app/blob/de9bcd7e5b82ac2b4fc80ec39d13d0c1bb6f6b9c/extensions/wikia/AdEngine/js/AdEngine2.js#L12): Function has 9 parameters which is greater than 7 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AExcessiveParameterList)
1. MAJOR [helper.js#L16](https://github.com/Wikia/app/blob/de9bcd7e5b82ac2b4fc80ec39d13d0c1bb6f6b9c/extensions/wikia/AdEngine/js/provider/gpt/helper.js#L16): Function has 12 parameters which is greater than 7 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AExcessiveParameterList)
1. MAJOR [helper.js#L51](https://github.com/Wikia/app/blob/de9bcd7e5b82ac2b4fc80ec39d13d0c1bb6f6b9c/extensions/wikia/AdEngine/js/provider/gpt/helper.js#L51): Function has a complexity of 14 which is greater than 10 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AFunctionComplexity)
1. INFO [AdEngine2.js#L58](https://github.com/Wikia/app/blob/de9bcd7e5b82ac2b4fc80ec39d13d0c1bb6f6b9c/extensions/wikia/AdEngine/js/AdEngine2.js#L58): Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)
1. INFO [AdEngine2.js#L81](https://github.com/Wikia/app/blob/de9bcd7e5b82ac2b4fc80ec39d13d0c1bb6f6b9c/extensions/wikia/AdEngine/js/AdEngine2.js#L81): Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 6 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [AdEngine2.js#L186](https://github.com/Wikia/app/blob/6ef34c72fb250b8f9b95453a6afe924b39986302/extensions/wikia/AdEngine/js/AdEngine2.js#L186): Change this condition so that it does not always evaluate to "true". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS2583)
1. MAJOR [AdEngine2.js#L12](https://github.com/Wikia/app/blob/6ef34c72fb250b8f9b95453a6afe924b39986302/extensions/wikia/AdEngine/js/AdEngine2.js#L12): Function has 9 parameters which is greater than 7 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AExcessiveParameterList)
1. MAJOR [helper.js#L16](https://github.com/Wikia/app/blob/6ef34c72fb250b8f9b95453a6afe924b39986302/extensions/wikia/AdEngine/js/provider/gpt/helper.js#L16): Function has 12 parameters which is greater than 7 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AExcessiveParameterList)
1. MAJOR [helper.js#L51](https://github.com/Wikia/app/blob/6ef34c72fb250b8f9b95453a6afe924b39986302/extensions/wikia/AdEngine/js/provider/gpt/helper.js#L51): Function has a complexity of 14 which is greater than 10 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AFunctionComplexity)
1. INFO [AdEngine2.js#L58](https://github.com/Wikia/app/blob/6ef34c72fb250b8f9b95453a6afe924b39986302/extensions/wikia/AdEngine/js/AdEngine2.js#L58): Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)
1. INFO [AdEngine2.js#L81](https://github.com/Wikia/app/blob/6ef34c72fb250b8f9b95453a6afe924b39986302/extensions/wikia/AdEngine/js/AdEngine2.js#L81): Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 7 issues
* CRITICAL 1 critical
* MAJOR 3 major
* MINOR 1 minor
* INFO 2 info

Watch the comments in this conversation to review them.

#### 6 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [AdEngine2.js#L186](https://github.com/Wikia/app/blob/15bb0a24b2aff7909400a7810c3ecacac3b37a60/extensions/wikia/AdEngine/js/AdEngine2.js#L186): Change this condition so that it does not always evaluate to "true". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS2583)
1. MAJOR [AdEngine2.js#L12](https://github.com/Wikia/app/blob/15bb0a24b2aff7909400a7810c3ecacac3b37a60/extensions/wikia/AdEngine/js/AdEngine2.js#L12): Function has 9 parameters which is greater than 7 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AExcessiveParameterList)
1. MAJOR [helper.js#L16](https://github.com/Wikia/app/blob/15bb0a24b2aff7909400a7810c3ecacac3b37a60/extensions/wikia/AdEngine/js/provider/gpt/helper.js#L16): Function has 12 parameters which is greater than 7 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AExcessiveParameterList)
1. MAJOR [helper.js#L51](https://github.com/Wikia/app/blob/15bb0a24b2aff7909400a7810c3ecacac3b37a60/extensions/wikia/AdEngine/js/provider/gpt/helper.js#L51): Function has a complexity of 14 which is greater than 10 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AFunctionComplexity)
1. INFO [AdEngine2.js#L58](https://github.com/Wikia/app/blob/15bb0a24b2aff7909400a7810c3ecacac3b37a60/extensions/wikia/AdEngine/js/AdEngine2.js#L58): Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)
1. INFO [AdEngine2.js#L81](https://github.com/Wikia/app/blob/15bb0a24b2aff7909400a7810c3ecacac3b37a60/extensions/wikia/AdEngine/js/AdEngine2.js#L81): Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 6 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [AdEngine2.js#L186](https://github.com/Wikia/app/blob/3be81d66b70b5c220b7f56da9f247733b16fc98f/extensions/wikia/AdEngine/js/AdEngine2.js#L186): Change this condition so that it does not always evaluate to "true". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS2583)
1. MAJOR [AdEngine2.js#L12](https://github.com/Wikia/app/blob/3be81d66b70b5c220b7f56da9f247733b16fc98f/extensions/wikia/AdEngine/js/AdEngine2.js#L12): Function has 9 parameters which is greater than 7 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AExcessiveParameterList)
1. MAJOR [helper.js#L16](https://github.com/Wikia/app/blob/3be81d66b70b5c220b7f56da9f247733b16fc98f/extensions/wikia/AdEngine/js/provider/gpt/helper.js#L16): Function has 12 parameters which is greater than 7 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AExcessiveParameterList)
1. MAJOR [helper.js#L51](https://github.com/Wikia/app/blob/3be81d66b70b5c220b7f56da9f247733b16fc98f/extensions/wikia/AdEngine/js/provider/gpt/helper.js#L51): Function has a complexity of 14 which is greater than 10 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AFunctionComplexity)
1. INFO [AdEngine2.js#L58](https://github.com/Wikia/app/blob/3be81d66b70b5c220b7f56da9f247733b16fc98f/extensions/wikia/AdEngine/js/AdEngine2.js#L58): Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)
1. INFO [AdEngine2.js#L81](https://github.com/Wikia/app/blob/3be81d66b70b5c220b7f56da9f247733b16fc98f/extensions/wikia/AdEngine/js/AdEngine2.js#L81): Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 6 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [AdEngine2.js#L186](https://github.com/Wikia/app/blob/efd91a1091d7a3a3fd65018ffa30fceb8e2776a1/extensions/wikia/AdEngine/js/AdEngine2.js#L186): Change this condition so that it does not always evaluate to "true". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS2583)
1. MAJOR [AdEngine2.js#L12](https://github.com/Wikia/app/blob/efd91a1091d7a3a3fd65018ffa30fceb8e2776a1/extensions/wikia/AdEngine/js/AdEngine2.js#L12): Function has 9 parameters which is greater than 7 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AExcessiveParameterList)
1. MAJOR [helper.js#L16](https://github.com/Wikia/app/blob/efd91a1091d7a3a3fd65018ffa30fceb8e2776a1/extensions/wikia/AdEngine/js/provider/gpt/helper.js#L16): Function has 12 parameters which is greater than 7 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AExcessiveParameterList)
1. MAJOR [helper.js#L51](https://github.com/Wikia/app/blob/efd91a1091d7a3a3fd65018ffa30fceb8e2776a1/extensions/wikia/AdEngine/js/provider/gpt/helper.js#L51): Function has a complexity of 14 which is greater than 10 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AFunctionComplexity)
1. INFO [AdEngine2.js#L58](https://github.com/Wikia/app/blob/efd91a1091d7a3a3fd65018ffa30fceb8e2776a1/extensions/wikia/AdEngine/js/AdEngine2.js#L58): Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)
1. INFO [AdEngine2.js#L81](https://github.com/Wikia/app/blob/efd91a1091d7a3a3fd65018ffa30fceb8e2776a1/extensions/wikia/AdEngine/js/AdEngine2.js#L81): Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 6 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [AdEngine2.js#L186](https://github.com/Wikia/app/blob/f5ea867d7b25d94046a4f842c1e53c7413950b65/extensions/wikia/AdEngine/js/AdEngine2.js#L186): Change this condition so that it does not always evaluate to "true". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS2583)
1. MAJOR [AdEngine2.js#L12](https://github.com/Wikia/app/blob/f5ea867d7b25d94046a4f842c1e53c7413950b65/extensions/wikia/AdEngine/js/AdEngine2.js#L12): Function has 9 parameters which is greater than 7 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AExcessiveParameterList)
1. MAJOR [helper.js#L16](https://github.com/Wikia/app/blob/f5ea867d7b25d94046a4f842c1e53c7413950b65/extensions/wikia/AdEngine/js/provider/gpt/helper.js#L16): Function has 12 parameters which is greater than 7 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AExcessiveParameterList)
1. MAJOR [helper.js#L51](https://github.com/Wikia/app/blob/f5ea867d7b25d94046a4f842c1e53c7413950b65/extensions/wikia/AdEngine/js/provider/gpt/helper.js#L51): Function has a complexity of 14 which is greater than 10 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AFunctionComplexity)
1. INFO [AdEngine2.js#L58](https://github.com/Wikia/app/blob/f5ea867d7b25d94046a4f842c1e53c7413950b65/extensions/wikia/AdEngine/js/AdEngine2.js#L58): Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)
1. INFO [AdEngine2.js#L81](https://github.com/Wikia/app/blob/f5ea867d7b25d94046a4f842c1e53c7413950b65/extensions/wikia/AdEngine/js/AdEngine2.js#L81): Complete the task associated to this TODO comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1135)

**Pull request merged**